File: coolrmi/hu.qgears.coolrmi/src/hu/qgears/coolrmi/multiplexer/SocketMultiplexer.java
package hu.qgears.coolrmi.multiplexer;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import hu.qgears.coolrmi.messages.AbstractCoolRMIMessage;



/**
 * Multiplexes several messages on a single TCP
 * connection to both direction.
 * 
 * Both write and read has an own thread.
 * 
 * @author rizsi
 *
 */
public class SocketMultiplexer {
	private ISocketMultiplexerListener messageListener;
	private InputStream is;
	private OutputStream os;
	private boolean exit=false;
	/**
	 * True means that the next messages can not be sent.
	 */
	private boolean disconnected;
	private LinkedList<SocketMultiplexerSource> messagesToSend=new LinkedList<SocketMultiplexerSource>();
	public SocketMultiplexer(InputStream is, OutputStream os,
			ISocketMultiplexerListener messageListener) throws IOException {
		super();
		this.is = is;
		this.messageListener=messageListener;
		this.os=os;
	}
	public void start()
	{
		new ReadThread().start();
		new WriteThread().start();
	}

	class ReadThread extends Thread
	{
		public ReadThread() {
			super("Cool RMI read thread");
		}

		@Override
		public void run() {
			try {
				try
				{
					while(!exit)
					{
						SocketMultiplexerDatagram datagram=SocketMultiplexerDatagram.readFromStream(is);
						messageListener.messageReceived(datagram.getContent());
					}
				}finally
				{
					is.close();
				}
			} catch (Exception e) {
				messageListener.pipeBroken(e);
			}
		}
	}

	class WriteThread extends Thread
	{
		public WriteThread() {
			super("Cool RMI write thread");
		}

		@Override
		public void run() {
			while(!exit)
			{
				SocketMultiplexerSource source = null;
				synchronized (messagesToSend) {
					if(messagesToSend.isEmpty())
					{
						try {
							messagesToSend.wait();
						} catch (InterruptedException e) {}
					}

					if (!messagesToSend.isEmpty()) {
						source = messagesToSend.removeFirst();
					}
				}
				if(source!=null)
				{
					SocketMultiplexerDatagram datagram=new SocketMultiplexerDatagram(source.getToSend());

					try {
						datagram.writeToStream(os);
						os.flush();
					} catch (IOException e) {
						synchronized (messagesToSend) {
							disconnected=true;
						}
						messageListener.pipeBroken(e);
					}

					source.sent();
				}
			}
		}
	}
	public void addMessageToSend(byte[] messageContent, AbstractCoolRMIMessage message)
	{
		boolean b;
		synchronized (messagesToSend) {
			messagesToSend.add(new SocketMultiplexerSource(messageContent, message));
			messagesToSend.notifyAll();
			b=disconnected;
		}
		if(b)
		{
			message.sent();
		}
	}
	public void stop()
	{
		exit=true;
		List<SocketMultiplexerSource> toCancel=null;
		synchronized (messagesToSend) {
			messagesToSend.notifyAll();
			toCancel=new ArrayList<SocketMultiplexerSource>(messagesToSend);
		}
		if(toCancel!=null)
		{
			for(SocketMultiplexerSource s: toCancel)
			{
				s.sent();
			}
		}
	}
}

File: coolrmi/CoolRMI.Net/Multiplexer/SocketMultiplexer.cs
using System;
using System.Collections.Concurrent;
using System.IO;
using System.Threading;
using CoolRMI.Net.Messages;

namespace CoolRMI.Net.Multiplexer
{
    public interface ISocketMultiplexerListener
    {
        void MessageReceived(byte[] bs);
        void PipeBroken(Exception e);
    }

    public class SocketMultiplexer
    {
        private readonly ISocketMultiplexerListener listener;
        private readonly Stream stream;
        private bool exit;

        private readonly BlockingCollection<SocketMultiplexerSource>
            messagesToSend = new BlockingCollection<SocketMultiplexerSource>();

        public SocketMultiplexer(Stream stream, ISocketMultiplexerListener listener)
        {
            this.listener = listener;
            this.stream = stream;
        }

        public void Start()
        {
            new Thread(ReadThread) {Name = "CoolRMI read thread"}.Start();
            new Thread(WriteThread) {Name = "CoolRMI write thread"}.Start();
        }

        private void ReadThread()
        {
            try
            {
                while (!exit)
                {
                    var datagram =
                        SocketMultiplexerDatagram.ReadFromStream(stream);
                    listener.MessageReceived(datagram.Content);
                }
            }
            catch (Exception e)
            {
                listener.PipeBroken(e);
            }
        }

        private void WriteThread()
        {
            SocketMultiplexerSource source;
            while (messagesToSend.TryTake(out source, -1))
            {
                var datagram = new SocketMultiplexerDatagram(source.ToSend);
                try
                {
                    datagram.WriteToStream(stream);
                    stream.Flush();
                }
                catch (IOException e)
                {
                    listener.PipeBroken(e);
                }
                source.Sent();
            }
        }

        public void AddMessageToSend(byte[] content,
            AbstractCoolRMIMessage message)
        {
            messagesToSend.Add(new SocketMultiplexerSource(content, message));
        }

        public void Stop()
        {
            exit = true;
            messagesToSend.CompleteAdding();

            SocketMultiplexerSource source;
            while (messagesToSend.TryTake(out source))
            {
                source.Sent();
            }
        }
    }
}
